File: androidApp/src/main/java/com/aglushkov/wordteacher/androidApp/features/di/DefinitionsModule.kt
package com.aglushkov.wordteacher.di

import com.aglushkov.wordteacher.androidApp.di.FragmentComp
import com.aglushkov.wordteacher.androidApp.features.definitions.blueprints.DefinitionsDisplayModeBlueprint
import com.aglushkov.wordteacher.androidApp.features.definitions.blueprints.WordDefinitionBlueprint
import com.aglushkov.wordteacher.androidApp.features.definitions.blueprints.WordDividerBlueprint
import com.aglushkov.wordteacher.androidApp.features.definitions.blueprints.WordExampleBlueprint
import com.aglushkov.wordteacher.androidApp.features.definitions.blueprints.WordPartOfSpeechBlueprint
import com.aglushkov.wordteacher.androidApp.features.definitions.blueprints.WordSubHeaderBlueprint
import com.aglushkov.wordteacher.androidApp.features.definitions.blueprints.WordSynonymBlueprint
import com.aglushkov.wordteacher.androidApp.features.definitions.blueprints.WordTitleBlueprint
import com.aglushkov.wordteacher.androidApp.features.definitions.blueprints.WordTranscriptionBlueprint
import com.aglushkov.wordteacher.androidApp.general.ViewItemBinder
import com.aglushkov.wordteacher.shared.features.definitions.repository.WordRepository
import com.aglushkov.wordteacher.shared.features.definitions.vm.DefinitionsVM
import com.aglushkov.wordteacher.shared.general.IdGenerator
import com.aglushkov.wordteacher.shared.general.connectivity.ConnectivityManager
import dagger.Module
import dagger.Provides

@Module
class DefinitionsModule {
    @FragmentComp
    @Provides
    fun createItemViewBinder(
        definitionsDisplayModeBlueprint: DefinitionsDisplayModeBlueprint
    ): ViewItemBinder {
        return ViewItemBinder()
            .addBlueprint(definitionsDisplayModeBlueprint)
            .addBlueprint(WordDefinitionBlueprint())
            .addBlueprint(WordDividerBlueprint())
            .addBlueprint(WordExampleBlueprint())
            .addBlueprint(WordPartOfSpeechBlueprint())
            .addBlueprint(WordSubHeaderBlueprint())
            .addBlueprint(WordSynonymBlueprint())
            .addBlueprint(WordTitleBlueprint())
            .addBlueprint(WordTranscriptionBlueprint())
    }

    @FragmentComp
    @Provides
    fun viewModel(
        connectivityManager: ConnectivityManager,
        wordRepository: WordRepository,
        idGenerator: IdGenerator,
        state: DefinitionsVM.State
    ): DefinitionsVM {
        return DefinitionsVM(connectivityManager, wordRepository, idGenerator, state)
    }
}

File: iosApp/iosApp/Features/Definitions/DI/DefinitionsModule.swift
//
//  DefinitionsModule.swift
//  iosApp
//
//  Created by Alexey Glushkov on 22.10.2020.
//  Copyright © 2020 orgName. All rights reserved.
//

import Foundation
import Cleanse
import shared

class DefinitionsModule: Module {
    typealias Scope = Singleton
    
    static func configure(binder: Binder<Singleton>) {
        binder.bind().sharedInScope().to(value: DefinitionsVM.State(word: nil) )
        binder.bind().sharedInScope().to(factory: DefinitionsVM.init.self)
        binder.bind().sharedInScope().to(factory: DefinitionsDisplayModeBlueprint.init.self)
        binder.bind().sharedInScope().to(factory: WordDefinitionBlueprint.init.self)
        binder.bind().sharedInScope().to(factory: WordDividerBlueprint.init.self)
        binder.bind().sharedInScope().to(factory: WordExampleBlueprint.init.self)
        binder.bind().sharedInScope().to(factory: WordPartOfSpeechBlueprint.init.self)
        binder.bind().sharedInScope().to(factory: WordSubHeaderBlueprint.init.self)
        binder.bind().sharedInScope().to(factory: WordSynonymBlueprint.init.self)
        binder.bind().sharedInScope().to(factory: WordTitleBlueprint.init.self)
        binder.bind().sharedInScope().to(factory: WordTranscriptionBlueprint.init.self)
        
        binder.bind().sharedInScope().to(factory: { (
                definitionsDisplayModeBlueprint: DefinitionsDisplayModeBlueprint,
                wordDefinitionBlueprint: WordDefinitionBlueprint,
                wordDividerBlueprint: WordDividerBlueprint,
                wordExampleBlueprint: WordExampleBlueprint,
                wordPartOfSpeechBlueprint: WordPartOfSpeechBlueprint,
                wordSubHeaderBlueprint: WordSubHeaderBlueprint,
                wordSynonymBlueprint: WordSynonymBlueprint,
                wordTitleBlueprint: WordTitleBlueprint,
                wordTranscriptionBlueprint: WordTranscriptionBlueprint
            ) in
            return ItemViewBinder()
                .addBlueprint(blueprint: definitionsDisplayModeBlueprint)
                .addBlueprint(blueprint: wordDefinitionBlueprint)
                .addBlueprint(blueprint: wordDividerBlueprint)
                .addBlueprint(blueprint: wordExampleBlueprint)
                .addBlueprint(blueprint: wordPartOfSpeechBlueprint)
                .addBlueprint(blueprint: wordSubHeaderBlueprint)
                .addBlueprint(blueprint: wordSynonymBlueprint)
                .addBlueprint(blueprint: wordTitleBlueprint)
                .addBlueprint(blueprint: wordTranscriptionBlueprint)
        })
    }
}
